### PR TITLE
feat: processTemplate inject variables and replace undefined variables with empty value, liv-4513

### PR DESCRIPTION
**Pull request by jinevo**

### CODE REVIEW

File: dist/io/processTemplate.js [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
+    // Inject the variables in a the global __VARIABLES__
+    const injectedVariables = `<script type="text/javascript">window.__VARIABLES__ = ${JSON.stringify(variables)}</script>`;
+    // Replace all occurences of {{ string }} with the variable (empty if not found)
+    const replacedOccurrences = content.replaceAll(/({{)([0-9a-zA-Z-\s]+)(}})/g, (fullMatch, match1, match2) => {
+        const variable = match2.trim();
+        const value = variable in variables ? variables[variable] : '';
+        if (typeof value === "object")
+            return JSON.stringify(value);
+        return String(value);
+    });
```
Comment (phsultan) [inline dist/io/processTemplate.js:13]: Since `JSON.stringify` can throw exception, we should probably wrap this in a try/catch and log so the dev knows we failed to parse their variables, what do you think?

File: tsconfig.json [hunk 2]
```diff
@@ -1,6 +1,7 @@
 {
   "compilerOptions": {
     "module": "commonjs",
+    "lib": ["DOM", "es2021"],
```
Comment (phsultan) [inline tsconfig.json:4]: Why is this needed?

File: package.json [hunk 3]
```diff
@@ -40,13 +40,14 @@
     "@types/uuid": "^8.3.0",
     "@typescript-eslint/eslint-plugin": "^4.13.0",
     "@typescript-eslint/parser": "^4.13.0",
+    "core-js": "^3.15.2",
```
Comment (phsultan) [inline package.json:43]: Why is this needed?

**Review by phsultan — COMMENTED** [review]
Great addition ! I love the testing part especially 👍 

I have a few questions regarding the APIs added, just for curiosity. Thanks!

File: tsconfig.json (hunk 2 above)
Comment (jinevo) [reply tsconfig.json:4]: The lib option allows to add/specify types libraries in the project.  Here i had to add "es2021" because I use the newly es2021 `String.protype.replaceAll` feature.

And since I've added lib, i had an DOM type error (`parent.postMessage(JSON.stringify(messageData), '*')` so i've added too.

File: package.json (hunk 3 above)
Comment (jinevo) [reply package.json:43]: It is needed for the tests to polyfill the `String.prototype.replaceAll` method (https://github.com/livestorm/livestorm-plugin/pull/16/files#diff-a1c786cdea90125d840669112d499d337a658bf6431a8c972a2e14301d908662R1)

Comment (phsultan) [reply package.json:43]: Not sure but I think you can call `String.prototype.replace` instead because the search argument is a regex with the `//g` flag. If so, this library should neither be needed.

Comment (jinevo) [reply package.json:43]: Indeed, it works with the `replace` method. `replaceAll` is better to use if the search arg is a string, this is not the case here so I will change to `replace` :)

File: dist/io/processTemplate.js (hunk 1 above)
Comment (jinevo) [reply dist/io/processTemplate.js]: Good idea !

File: yarn.lock [hunk 4]
```diff
@@ -4446,9 +4446,9 @@ typedarray-to-buffer@^3.1.5:
     is-typedarray "^1.0.0"
 
 typescript@^4.1.3:
-  version "4.1.3"
-  resolved "https://registry.yarnpkg.com/typescript/-/typescript-4.1.3.tgz#519d582bd94cba0cf8934c7d8e8467e473f53bb7"
-  integrity sha512-B3ZIOf1IKeH2ixgHhj6la6xdwR9QrLC5d1VKeCSY4tvkqhF2eqd9O7txNlS0PO3GrBAFIdr3L1ndNwteUbZLYg==
+  version "4.3.5"
```
Comment (phsultan) [inline yarn.lock:4449]: The `yarn.lock` file should be regenerated since you restored Typescript to `4.1.3`.

Comment (jinevo) [reply yarn.lock:4449]: I've already re-generated it. With the semver rule, it has upgraded to 4.3.5. I've left it because there are no changes in the built files due to this upgrade.

Comment (phsultan) [reply yarn.lock:4449]: 👍

**Review by phsultan — APPROVED** [review]

**Review by Michaelvilleneuve — APPROVED** [review]
